### ISSUE TITLE
jitsimeet ipad camera rotation

### ISSUE THREAD

**Issue by ysbutuner** [jitsi/jitsi-meet#14111]
In jitsi meet iPad devices, the camera image does not rotate on its own screen in horizontal and vertical orientation, but it is displayed correctly to the participants on the opposite side.

**Comment by saghul** [comment]
Can you share a screenshot of the problem?

**Comment by saghul** [comment]
Ah got it. This is a known issue with ipads. We are working to solve it. IN the meantime rotating the device twice should reset it.

**Comment by ysbutuner** [comment]
I understand, can you let me know when you solve it? I constantly encounter this bug.

thank you for your attention

**Comment by saghul** [comment]
Don't hold your breadth, it requires patching WebRTC and I tried once but the patch got rejected so I'll need to take another look.

**Comment by ysbutuner** [comment]
I understand, could you share with me the change you made in WEBRTC, I would like to add the change and see if the problem is solved.

**Comment by ysbutuner** [comment]
Can you provide support on this matter? @saghul

**Comment by ysbutuner** [comment]
Yes, I am ready, but I could not find the source of the problem in the code, I can do it if you support me.

**Comment by saghul** [comment]
I'll find the link.

**Comment by ysbutuner** [comment]
Thank you, I'm waiting for your return

**Comment by ysbutuner** [comment]
I will thank you for your interest, I will work on this issue and keep you informed.

**Comment by lesinigo** [comment]
We've been hitten by the same bug developing a custom iPad application that uses the Jitsi SDK.

@saghul thanks for the work on this issue on WebRTC's side. Afaict in the SDK you're actually using a build of WebRTC from [your own mirror](https://github.com/jitsi/webrtc) and with additional patches on top of the upstream version: would you consider to add your proposed WebRTC patch to your mirror's branch?

If that can't work out: is there an older, known good version of the Jitsi SDK that's not affected by this bug (possibly because it's using an older WebRTC build) ? By reading all the material here and on the WebRTC side I sill cannot pinpoint where the defect had initially appeared.

**Comment by lesinigo** [comment]
We took a look at it but we really aren't familiar with the tools used by the jitsi/webrtc build script.

Could you build and upload the `WebRTC.xcframework.zip` of the patched version somewhere? That would enable us to fork `jitsi/jitsi-meet-ios-sdk-releases`, point it at the patched WebRTC, build our app and report feedback on how it behaves... maybe that could be useful for reporting to upstream too, I see they asked for unit tests and are not moving forward from that.

**Comment by saghul** [comment]
Sorry I can't, it takes time and alas I don't have it now.

In terms of feedback, it did work for me, what's missing is to give it another try and see if the reviewer was right and a new approach is necessary. That is not something that testing as an end user will yield.

**Comment by lesinigo** [comment]
In the past months I had the pleasure to get to know how to recompile `WebRTC.xcframework.zip` and even automated it in our internal CI. This allowed us to rapidly iterate over the proposed patch and some variants thereof, but none have had any effect on the camera rotation bug, it only made it even worse.

As far as web can tell other people connected to the same room never see any camera rotation problem, the wrong orientation seems to affect only the screen of the device that's the source of the video stream. In other words, I see myself rotated but everyone else is seeing me correctly and I see everyone else correctly.

When trying to hardcode a fixed `UIDeviceOrientation*` value in the patch (instead of actually reacting to different device orientations) we got the effect of having connected people seeing me rotated, while not fixing the bug of me seeing myself rotated. Additionally, when taking an actual screenshot it comes out correct, we had to photograph the iPad using another smartphone to get pictures of the defect.

This leads us to suspect that the bug lies in the code path between the camera and the display (some kind of hardware acceleration stuff that's not involved at screenshot time?), not between the camera and the network (because other people always see me correctly). I don't know if this means that it's on WebRTC side or on Jitsi Meet's side, but I can relate that to [the WebRTC maintainer question of having some further test cases](https://webrtc-review.googlesource.com/c/src/+/311420/1#message-bd322d99a37513374bfdb572e89b85d4bad4ba01) before considering the patch.

I'm trying to contact some involved parties directly (including 8x8 Inc) to try and find out if someone is willing to take ownership of this issue and fix it once and for all.

**Comment by soniccrash7** [comment]
Hi, is there any news regarding this problem? We are also experiencing it on newer iPads starting with Ipad's machine version: iPad8,1 : iPad Pro 11 inch 3rd Gen (WiFi)

When the meeting is in portrait everything is ok but when rotated in landscape the self video is rotated always sideways. The other participants sees me correctly.

This can also be reproduced on the official Jitsi Meet app running on Ipad 9th gen

We tried with custom build from the JitsiWebRTC and transformed the RTCMTLVIdeoView by applying CGAffineTransformScaleMake(-1.0, 1.0), and while it resolves the issue on all Ipads it freezes the video when switching between tile and single view in landscape after some random time. This might be causing some overhead and it's causing the freeze.
